flake8_trio.py

```diff
@@ -34,29 +33,68 @@ def is_trio_call(node: ast.AST, *names: str) -> Optional[str]:
 class Visitor(ast.NodeVisitor):
     def __init__(self) -> None:
         super().__init__()
-        self.problems: list[tuple[int, int]] = []
+        self.problems: List[Error] = []
 
     def visit_With(self, node: ast.With) -> None:
-        for item in node.items:
-            call = is_trio_call(item.context_expr, "fail_after", "move_on_after")
-            if call and not any(isinstance(x, ast.Await) for x in ast.walk(node)):
-                self.problems.append(
-                    (item.lineno, item.col_offset, TRIO100.format(call))
-                )
+        self.check_for_trio100(node)
+        self.generic_visit(node)
 
-        # Don't forget to visit the child nodes for other errors!
+    def visit_AsyncWith(self, node: ast.AsyncWith) -> None:
+        self.check_for_trio100(node)
         self.generic_visit(node)
 
+    def check_for_trio100(self, node: Union[ast.With, ast.AsyncWith]) -> None:
+        # Context manager with no `await` call within
+        for item in (i.context_expr for i in node.items):
+            call = is_trio_call(item, "fail_after", "move_on_after")
+            if call and not any(isinstance(x, ast.Await) for x in ast.walk(node)):
+                self.problems.append(TRIO100(item.lineno, item.col_offset, call))
+
 
 class Plugin:
     name = __name__
     version = __version__
 
-    def __init__(self, tree: ast.AST) -> None:
-        self._tree = tree
+    def __init__(
+        self, tree: Optional[ast.AST] = None, filename: Optional[str] = None
+    ) -> None:
+        if tree is None:
+            assert filename is not None
+            self._tree = self.load_file(filename)
+        else:
+            self._tree = tree
+
+    def load_file(self, filename: str) -> ast.AST:
+        """Loads the file in a way that auto-detects source encoding.
 
-    def run(self) -> Generator[tuple[int, int, str, Type[Any]], None, None]:
+        Stolen from flake8_import_order because it's good.
+        """
+        lines = pycodestyle.readlines(filename)
+        return ast.parse("".join(lines))
```

Comment:
```suggestion
        with tokenize.open(filename) as f:
            return ast.parse(f.read())
```

I'm not actually a fan of the "on error assume UTF-8" approach; it seems likely to cause syntax problems later and doesn't actually fix the encoding error.

Reply:
This one I more or less just straight copy-pasted w/o much investigation. Sounds good, and yeah it looked quite ugly after I skipped out on `stdin` support.